stress-gpu.c

```diff
@@ -0,0 +1,414 @@
+    "precision mediump float;\n"
+    "varying vec4 v_color;\n"
+    "uniform int frag_n;\n"
+    "void main()\n"
+    "{\n"
+    "int i;\n"
+    "vec4 a = v_color;\n"
+    "for (i = 0; i < frag_n; i++) {\n"
+    "float f = float(i);\n"
+    "a = a / clamp(sin(f) * exp(f), 0.1, 0.9);\n"
+    "}\n"
+    "a = clamp(a, -1.0, 1.0);\n"
+    "gl_FragColor = v_color + 0.000001 * a;\n" "}\n";
+
+static GLuint compile_shader(const char *text, int size, GLenum type)
+{
+	GLuint shader;
+	GLint compiled;
+	const GLchar *source[1] = { text };
+
+	shader = glCreateShader(type);
+	glShaderSource(shader, 1, source, &size);
+	glCompileShader(shader);
+	glGetShaderiv(shader, GL_COMPILE_STATUS, &compiled);
+	if (!compiled) {
+		GLint infoLen = 0;
+		glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &infoLen);
+		if (infoLen > 1) {
+			char *infoLog = malloc(infoLen);
+			glGetShaderInfoLog(shader, infoLen, NULL, infoLog);
+			(void)fprintf(stderr, "Error compiling shader:\n%s\n",
+				      infoLog);
+			free(infoLog);
+		}
+		glDeleteShader(shader);
+		return 0;
+	}
+	return shader;
+}
+
+void load_shaders(void)
+{
+	GLint linked;
+	GLuint vertexShader;
+	GLuint fragmentShader;
+	if ((vertexShader =
+	     compile_shader(vert_shader, sizeof(vert_shader),
+			    GL_VERTEX_SHADER)) == 0)
+		(void)fprintf(stderr, "Failed to compile vertex shader\n");
+	if ((fragmentShader =
+	     compile_shader(frag_shader, sizeof(frag_shader),
+			    GL_FRAGMENT_SHADER)) == 0)
+		(void)fprintf(stderr, "Failed to compile fragment shader\n");
+	if ((program = glCreateProgram()) == 0)
+		(void)fprintf(stderr, "Error creating the shader program\n");
+	glAttachShader(program, vertexShader);
+	glAttachShader(program, fragmentShader);
+	glLinkProgram(program);
+	glGetProgramiv(program, GL_LINK_STATUS, &linked);
+	if (!linked) {
+		GLint infoLen = 0;
+		glGetProgramiv(program, GL_INFO_LOG_LENGTH, &infoLen);
+		if (infoLen > 1) {
+			char *infoLog = malloc(infoLen);
+			glGetProgramInfoLog(program, infoLen, NULL, infoLog);
+			(void)fprintf(stderr,
+				      "Error linking shader program:\n%s\n",
+				      infoLog);
+			free(infoLog);
+		}
+		glDeleteProgram(program);
+		exit(EXIT_FAILURE);
+	}
+
+	glUseProgram(program);
+}
+
+static const GLfloat vertex[] = {
+	-1, -1, 0, 1,
+	-1, 1, 0, 1,
+	1, 1, 0, 1,
+
+	-1, -1, 0, 1,
+	1, 1, 0, 1,
+	1, -1, 0, 1,
+};
+
+static const GLfloat color[] = {
+	1, 0, 0, 1,
+	0, 1, 0, 1,
+	0, 0, 1, 1,
+
+	1, 0, 0, 1,
+	0, 0, 1, 1,
+	1, 1, 0, 1,
+};
+
+void gles2_init(uint32_t width, uint32_t height, int frag_n, GLsizei texsize)
+{
+	pr_inf("GL_VENDOR: %s\n", (char *)glGetString(GL_VENDOR));
+	pr_inf("GL_VERSION: %s\n", (char *)glGetString(GL_VERSION));
+	pr_inf("GL_RENDERER: %s\n", (char *)glGetString(GL_RENDERER));
+
+	load_shaders();
+
+	glClearColor(0, 0, 0, 0);
+	glViewport(0, 0, width, height);
+
+	GLint ufrag_n = glGetUniformLocation(program, "frag_n");
+	glUniform1i(ufrag_n, frag_n);
+	if (glGetError() != GL_NO_ERROR)
+		(void)fprintf(stderr,
```

Comment:
I see stress-ng has some log helpers like pr_err (pf_inf is used above), maybe those could be used overall in this new file too. Probably not strictly required though as not all tests do this.

Reply:
ok, using pr_inf then